### ISSUE TITLE
Checking matching begin / end entries

### ISSUE THREAD

**Issue by mikael-s-persson** [schulmar/Templar#5]
This is kind of an issue with both templar and templight. I recently tried to load a very large templight profile with templar. The attempt failed on a seg-fault because there seems to be a mismatching of begin and end entries in the templight trace file.

The templar file-reading code makes the assumption that every time `TraceReader::endEntry()` is called, the "end" entry is the end of the "begin" entry that is currently on top of the stack `childVectorStack`. This is certainly a reasonable assumption, but it seems that there are rare cases where there might be an "end" entry with no matching "begin" entry or vice versa. In those cases, that whole code breaks down, of course (and ends up with a seg-fault or just a corrupt graph).

I'm not exactly sure where this mismatch is coming from. It could be coming from a bug in templight, or it could be coming from a specific set of circumstances that make Clang call templight in a mismatching way (i.e., the problem is in Clang, not templight).

Either way, I can try to fix it, but I cannot guarantee that such things will never occur. So, I'm thinking that it would be best to have a correction mechanism in the traces. If there was a way to verify that a given "end" entry matches a given "begin" entry (like an id number or some other common information), than there would at least be a way to repair / recover / ignore mismatching entries and limiting the damage to only those few entries. What do you think?

**Comment by mikael-s-persson** [comment]
Yeah, I agree with all that you've said.

> 1) try to solve this in templight/clang

I did. I found that the problem was much bigger than I original thought, and the bug was in the logic for skipping entries (ignored system templates, or blacklisted templates) which cause some trailing end-blocks to get through. Anyways, it's fixed (for now!).

> offer a sanitizer

Actually, as I was thinking of how to create an id field that would be best for sanitizing afterwards, it boiled down to doing some kind of a hash of the instantiation-kind and the AST entity pointer, which are the only two fields that match in both the begin and end entries (as templight gets them from clang) and that are sufficient to uniquely verify the match. And as I realized that, I realized that the best place to sanitize the output was in templight itself, when that information is available. So, with the current chain "clang -1-> templight tracer -2-> templight file-writer -3-> templar", I cannot really guarantee anything about link "-1->" because clang could be buggy and it's beyond my control, but I can guarantee a sane output in link "-2->" because I have the necessary information to verify it, and that is what I have done with my latest commit (it adds a bit of overhead to the templight tracer, but it's probably worth the cost). And the file writers (within templight or templight-convert, or third-party converters) and readers (e.g., templar, or templight-convert, etc.) are a lot easier to write correctly and therefore provide some reasonable expectation that they work correctly.

If a similar problems pops up again, I'll consider adding an optional unique id field to each begin-end pair, so that a separate sanitizer (or checker) is at least possible. But I'll leave it as is for now.
